src/linux/component.rs

```diff
@@ -1,160 +1,369 @@
+    /// temp[1-*]_min Temperature min value.
+    ///     Unit: millidegree Celsius
+    ///     RW
+    threshold_min: Option<f32>,
+    /// temp[1-*]_crit	Temperature critical max value, typically greater than
+    /// 	corresponding temp_max values.
+    /// 	Unit: millidegree Celsius
+    /// 	RW
+    threshold_critical: Option<f32>,
+    /// Rare to have that value
+    /// temp[1-*]_type	Sensor type selection.
+    ///	Integers 1 to 6
+    ///	RW
+    ///	1: CPU embedded diode
+    ///	2: 3904 transistor
+    ///	3: thermal diode
+    ///	4: thermistor
+    ///	5: AMD AMDSI
+    ///	6: Intel PECI
+    ///	Not all types are supported by all chips
+    sensor_type: Option<TermalSensorType>,
+    /// May be empty!
+    /// temp[1-*]_label	Suggested temperature channel label.
+    /// Text string
+    /// Should only be created if the driver has hints about what
+    /// this temperature channel is being used for, and user-space
+    /// doesn't. In all other cases, the label is provided by
+    /// user-space.
+    /// RO
     label: String,
+    // TODO:
+    // temp[1-*]_lowest
+    // Historical minimum temperature
+    // Unit: millidegree Celsius
+    // RO
+    //
+    // temp[1-*]_highest
+    // Historical maximum temperature
+    // Unit: millidegree Celsius
+    //
+    // temp[1-*]_lcrit	Temperature critical min value, typically lower than
+    // corresponding temp_min values.
+    // Unit: millidegree Celsius
+    // RW
+    //
+    // temp[1-*]_emergency
+    // 	Temperature emergency max value, for chips supporting more than
+    // 	two upper temperature limits. Must be equal or greater than
+    // 	corresponding temp_crit values.
+    // 	Unit: millidegree Celsius
+    // 	RW
     input_file: PathBuf,
+    highest_file: PathBuf,
 }
 
+// Read arbitrary data from sysfs.
 fn get_file_line(file: &Path, capacity: usize) -> Option<String> {
     let mut reader = String::with_capacity(capacity);
-    if let Ok(mut f) = File::open(file) {
-        if f.read_to_string(&mut reader).is_ok() {
-            Some(reader)
-        } else {
-            None
-        }
-    } else {
-        None
-    }
+    let mut f = File::open(file).ok()?;
+    f.read_to_string(&mut reader).ok()?;
+    reader.truncate(reader.trim_end().len());
+    Some(reader)
+}
+
+/// Read an `i32` aka `c_long` from a `/sys/class/hwmon` sysfs file.
+#[inline]
+fn get_i32_from_file(file: &Path) -> Option<i32> {
+    let mut reader = [0u8; 32];
+    let mut f = File::open(file).ok()?;
+    let end = f.read(&mut reader).ok()?;
+    let number = std::str::from_utf8(&reader[..end]).ok()?;
+    let number = number.trim();
+    i32::from_str_radix(number, 10).ok()
+}
+
+#[inline]
+fn get_u32_from_file(file: &Path) -> Option<u32> {
+    let mut reader = [0u8; 32];
+    let mut f = File::open(file).ok()?;
+    f.read(&mut reader).ok()?;
+    let number = std::str::from_utf8(&reader).ok()?;
+    //let number = number.trim();
+    u32::from_str_radix(number, 10).ok()
+}
```

Comment:
Instead of having two functions, I think it'd be simpler to have:

```suggestion
#[inline]
fn get_number_from_file<T>(file: &Path) -> Option<T> {
    let mut reader = [0u8; 32];
    let mut f = File::open(file).ok()?;
    f.read(&mut reader).ok()?;
    let number = std::str::from_utf8(&reader).ok()?;
    //let number = number.trim();
    T::from_str_radix(number, 10).ok()
}
```

I don't think it'll work as is, it might need to create a trait which would call `from_str_radix`. If that seems too annoying, you could instead do an inner call like this:

```suggestion
fn get_i32_from_file(file: &Path) -> Option<i32> {
    i32::from_str_radix(get_number_inner()?.trim(), 10).ok()
}

#[inline]
fn get_u32_from_file(file: &Path) -> Option<u32> {
    u32::from_str_radix(get_number_inner()?.trim(), 10).ok()
}

#[inline]
fn get_number_inner(file: &Path) -> Option<String> {
    let mut reader = [0u8; 32];
    let mut f = File::open(file).ok()?;
    f.read(&mut reader).ok()?;
    std::str::from_utf8(&reader).ok()
}
```

Reply:
No problem with that, I will check it out tonight, tried it don't remember why I went to the full-spell it out version.

Reply:
Think I have done something, I remember why I used `u32::from_str_radix` it don't bother about `'0` XD

Reply:
It's done, suppressed the comment said something stupid (wanted not being indexed since it was bad explanation) just had a function misbehaving.